packages/esm-amd-loader/src/esm-amd-loader.ts

```diff
@@ -18,42 +18,349 @@ interface Window {
+      });
+      continue;
+    }
+
+    // We have a dependency on a real module.
+    const dependency = getModule(resolveUrl(module.urlBase, depSpec));
+    args.push(dependency.exports);
+    depModules.push(dependency);
+
+    if (dependency.stateData.state === StateEnum.Initialized) {
+      load(dependency as Module<Initialized>);
+    }
+  }
+  return [args, depModules];
+}
+
+/**
+ * Start executing our dependencies, in order, as they become available.
+ * Once they're all executed, execute our own module body, if any.
+ */
+function beginWaitingOnDeps(module: Module<WaitingOnEarlierScripts>) {
+  const stateData: WaitingOnDeps = {
+    state: StateEnum.WaitingOnDeps,
+    args: module.stateData.args,
+    deps: module.stateData.deps,
+    moduleBody: module.stateData.moduleBody
+  };
+  const mutatedModule = stateTransition(module, stateData);
+  waitOnDeps(
+      module.stateData.deps,
+      () => execute(mutatedModule),
+      (e) => fail(mutatedModule, e));
+  return mutatedModule;
+}
+
+/** Runs the given module body. */
+function execute(module: Module<WaitingOnDeps>): Module<Executed|Failed> {
+  const stateData = module.stateData;
+  if (stateData.moduleBody != null) {
+    try {
+      stateData.moduleBody.apply(null, stateData.args);
+    } catch (e) {
+      return fail(module, e);
+    }
+  }
+  return stateTransition(module, {state: StateEnum.Executed});
+}
+
+/**
+ * Called when a module has failed to load, either becuase its script errored,
+ * or because one of its transitive dependencies errored.
+ */
+function fail(mod: Module, error: Error) {
+  return stateTransition(mod, {state: StateEnum.Failed, error});
+}
+
+/**
+ * Transition the given module to the given state.
+ *
+ * Does not do any checking that the transition is legal.
+ * Calls onNextStateChange callbacks.
+ */
+function stateTransition<SD extends ModuleStateData>(
+    module: Module, newStateData: SD): Module<SD> {
+  if (debugging) {
+    console.log(`${module.url} transitioning to state ${newStateData.state}`);
+  }
+  const mutatedModule = module as Module<SD>;
+  mutatedModule.stateData = newStateData;
+  if (mutatedModule.onNextStateChange.length > 0) {
+    const callbacks = mutatedModule.onNextStateChange.slice();
+    mutatedModule.onNextStateChange.length = 0;
+    for (const callback of callbacks) {
+      callback();
+    }
+  }
+  return mutatedModule;
+}
+
+function waitOnDeps(
+    deps: Module[],
+    onResolve: ResolveCallback|undefined,
+    onError: ErrorCallback|undefined): void {
+  const nextDep = deps.shift();
+  if (nextDep === undefined) {
+    // all deps resolved
+    if (onResolve) {
+      onResolve();
+    }
+    return;
+  }
+
+  if (nextDep.stateData.state === StateEnum.WaitingOnDeps) {
+    if (debugging) {
+      console.log(`Cycle detected importing ${nextDep.url}`);
+    }
+    // Do not wait on the dep that introduces a cycle, press on as though it
+    // were not there.
+    waitOnDeps(deps, onResolve, onError);
```

Comment:
should there be a `return` here? seems like you still end up waiting on the dep right now.

Reply:
Yep, that's a bug!